userspace/libscap/scap_savefile.c

```diff
@@ -2446,81 +2734,168 @@ int32_t scap_next_offline(scap_t *handle, OUT scap_evt **pevent, OUT uint16_t *p
 		{
-			//
-			// We read exactly 0 bytes. This indicates a correct end of file.
-			//
-			return SCAP_EOF;
+			handle->m_last_evt_dump_flags = *(uint32_t*)(handle->m_file_evt_buf + sizeof(uint16_t));
+			*pevent = (struct ppm_evt_hdr *)(handle->m_file_evt_buf + sizeof(uint16_t) + sizeof(uint32_t));
 		}
 		else
 		{
-			CHECK_READ_SIZE(readsize, sizeof(bh));
+			handle->m_last_evt_dump_flags = 0;
+			*pevent = (struct ppm_evt_hdr *)(handle->m_file_evt_buf + sizeof(uint16_t));
 		}
-	}
 
-	if(bh.block_type != EV_BLOCK_TYPE &&
-		bh.block_type != EV_BLOCK_TYPE_INT &&
-		bh.block_type != EVF_BLOCK_TYPE)
-	{
-		snprintf(handle->m_lasterr, SCAP_LASTERR_SIZE, "unexpected block type %u", (uint32_t)bh.block_type);
-		handle->m_unexpected_block_readsize = readsize;
-		return SCAP_UNEXPECTED_BLOCK;
-	}
-
-	if(bh.block_total_length < sizeof(bh) + sizeof(struct ppm_evt_hdr) + 4)
-	{
-		snprintf(handle->m_lasterr, SCAP_LASTERR_SIZE, "block length too short %u", (uint32_t)bh.block_total_length);
-		return SCAP_FAILURE;
-	}
+		if((*pevent)->type >= PPM_EVENT_MAX)
+		{
+			//
+			// We're reading a capture that contains new syscalls.
+			// We can't do anything else that skips them.
+			//
+			continue;
+		}
 
-	//
-	// Read the event
-	//
-	readlen = bh.block_total_length - sizeof(bh);
-	if (readlen > FILE_READ_BUF_SIZE) {
-		snprintf(handle->m_lasterr, SCAP_LASTERR_SIZE, "event block length %u greater than read buffer size %u",
-			 readlen,
-			 FILE_READ_BUF_SIZE);
-		return SCAP_FAILURE;
-	}
+		if(bh.block_type != EV_BLOCK_TYPE_V2 && bh.block_type != EVF_BLOCK_TYPE_V2)
+		{
+			//
+			// We're reading a old capture which events don't have nparams in the header.
+			// Convert it to the current version.
+			//
+			if((readlen + sizeof(uint32_t)) > FILE_READ_BUF_SIZE)
+			{
+				snprintf(handle->m_lasterr, SCAP_LASTERR_SIZE, "cannot convert v1 event block to v2 (%lu greater than read buffer size %u)",
+					 readlen + sizeof(uint32_t),
+					 FILE_READ_BUF_SIZE);
+				return SCAP_FAILURE;
+			}
 
-	readsize = gzread(f, handle->m_file_evt_buf, readlen);
-	CHECK_READ_SIZE(readsize, readlen);
+			memmove((char *)*pevent + sizeof(struct ppm_evt_hdr),
+				(char *)*pevent + sizeof(struct ppm_evt_hdr) - sizeof(uint32_t),
+				readlen - ((char *)*pevent - handle->m_file_evt_buf) - (sizeof(struct ppm_evt_hdr) - sizeof(uint32_t)));
+			(*pevent)->len += sizeof(uint32_t);
 
-	//
-	// EVF_BLOCK_TYPE has 32 bits of flags
-	//
-	*pcpuid = *(uint16_t *)handle->m_file_evt_buf;
+			//
+			// The number of parameters needs to be calculated based on the block len.
+			// Use the current number of parameters as starting point and decrease it
+			// until size matches.
+			//
+			char *end = (char *)*pevent + (*pevent)->len;
+			uint16_t *lens = (uint16_t *)((char *)*pevent + sizeof(struct ppm_evt_hdr));
+			uint32_t nparams;
+			bool done = false;
+			for(nparams = g_event_info[(*pevent)->type].nparams; nparams >= 0; nparams--)

```

Comment:
`nparams` is `uint32`, so it doesn't work well with a `nparams >= 0`. I'm sure it doesn't cause too much troubles in practice, but it's still a bad check.

Reply:
`nparams` is a `uint32_t` because `ppm_event_info.nparams` is of that type.
I'll cast it in the check.